### PR TITLE
Implement ToConstraintFieldGadget for Vec<UInt8>

### PR DESCRIPTION
**Pull request by Will-Lin4**

### CODE REVIEW

File: src/bits/uint8.rs [hunk 1]
```diff
@@ -334,12 +335,27 @@ impl<ConstraintF: Field> AllocVar<u8, ConstraintF> for UInt8<ConstraintF> {
...
+// Assumes bytes are little endian
+impl<ConstraintF: PrimeField> ToConstraintFieldGadget<ConstraintF> for Vec<UInt8<ConstraintF>> {
+    #[tracing::instrument(target = "r1cs")]
+    fn to_constraint_field(&self) -> Result<Vec<FpVar<ConstraintF>>, SynthesisError> {
+        let max_size = (ConstraintF::Params::CAPACITY / 8) as usize;
+        self.chunks(max_size)
+            .map(|chunk| Boolean::le_bits_to_fp_var(chunk.to_bits_le()?.as_slice()))
+            .collect::<Result<Vec<_>, SynthesisError>>()
+    }
+}
```
Comment (Pratyush) [inline src/bits/uint8.rs:347]: Maybe we should add pointers to the `ToConstraintField` impl for `Vec<UInt8>`?

Comment (weikengchen) [reply src/bits/uint8.rs:347]: Pratyush, do you mean &[UInt8<>]?

Comment (weikengchen) [reply src/bits/uint8.rs:347]: or something like this:
```

impl<'a, F: Field, T: 'a + R1CSVar<F>> R1CSVar<F> for &'a T {
    type Value = T::Value;

    fn cs(&self) -> ark_relations::r1cs::ConstraintSystemRef<F> {
        (*self).cs()
    }

    fn value(&self) -> Result<Self::Value, ark_relations::r1cs::SynthesisError> {
        (*self).value()
    }
}
```

**Review by weikengchen — APPROVED** [review]
Code looks good.

File: src/bits/uint8.rs (hunk 1 above)
Comment (Will-Lin4) [reply src/bits/uint8.rs:347]: I don't fully understand what you mean by adding pointers. Is it what Weikeng is referring to above?

Comment (ValarDragon) [reply src/bits/uint8.rs:347]: In the native case, I feel like if there was method named `Vec<u8>.to_field()` I'd expect it to return a single field element (the equivalent of `from_bytes(Vec<u8>)`. I think its confusing that this instead casts each individual byte.

Comment (Pratyush) [reply src/bits/uint8.rs:347]: @Will-Lin4 i meant just adding a link to the corresponding native impl on [u8] (to justify constants and such)

Comment (Pratyush) [reply src/bits/uint8.rs:347]: @valardragon the corresponding native code doesn’t cast each byte to a field element, but rather divides the stream of bytes into chunks of bytes, such that each chunk, when interpreted as an integer, is less than F::MODULUS

Comment (ValarDragon) [reply src/bits/uint8.rs:347]: oh I see, I misread it. That sounds good

Comment (ValarDragon) [reply src/bits/uint8.rs:347]: Lets add a comment here explaining that this is what its doing?

File: src/bits/uint8.rs [hunk 2]
```diff
@@ -334,12 +335,37 @@ impl<ConstraintF: Field> AllocVar<u8, ConstraintF> for UInt8<ConstraintF> {
     }
 }
 
+/// Converts `Vec<UInt8<ConstraintF>>`, which is assumed to be little-endian, to its
+/// `Vec<FpVar<ConstraintF>>` representation.
```
Comment (ValarDragon) [inline src/bits/uint8.rs:339]: Could this add a sentence that it parses the bytes in fixed size `ConstraintF::Params::CAPACITY` chunks?

**Review by ValarDragon — APPROVED** [review]
LGTM, beyond my last request for a comment

**Pull request merged**